web/text_layer_builder.js

```diff
@@ -83,37 +90,82 @@ class TextLayerBuilder {
    * @param {number} [timeout] - Wait for a specified amount of milliseconds
    *                             before rendering.
    */
-  render(timeout = 0) {
-    if (!(this.textContent || this.textContentStream) || this.renderingDone) {
+  render(viewport, timeout = 0) {
+    if (!(this.textContent || this.textContentStream)) {
       return;
     }
-    this.cancel();
 
-    this.textDivs.length = 0;
-    this.highlighter?.setTextMapping(this.textDivs, this.textContentItemsStr);
-    this.accessibilityManager?.setTextMapping(this.textDivs);
-
-    const textLayerFrag = document.createDocumentFragment();
-    this.textLayerRenderTask = renderTextLayer({
-      textContent: this.textContent,
-      textContentStream: this.textContentStream,
-      container: textLayerFrag,
-      viewport: this.viewport,
-      textDivs: this.textDivs,
-      textContentItemsStr: this.textContentItemsStr,
-      timeout,
-    });
-    this.textLayerRenderTask.promise.then(
-      () => {
-        this.textLayerDiv.append(textLayerFrag);
-        this.#finishRendering();
-        this.highlighter?.enable();
-        this.accessibilityManager?.enable();
-      },
-      function (reason) {
-        // Cancelled or failed to render text layer; skipping errors.
+    const scale = viewport.scale * (globalThis.devicePixelRatio || 1);
+
+    if (this.renderingDone) {
+      this.hide();
+      this.#update(viewport);
+      if (scale !== this.scale) {
+        updateTextLayer({
+          viewport,
+          scale,
+          textDivs: this.textDivs,
+          textDivProperties: this.textDivProperties,
+          isOffscreenCanvasSupported: this.isOffscreenCanvasSupported,
+        });
```

Comment:
Should we not dispatch a "textlayerrendered" event in this case as well, since otherwise there might be functionality (even in the viewer itself) which no longer works correctly!?